src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs

```diff
@@ -170,5 +177,95 @@ public RestClient FindRestClient(OperationGroup operationGroup)
             return model;
         }
+
+        private void DecorateOperationGroup()
+        {
+            foreach (var operations in _codeModel.OperationGroups)
+            {
+                operations.ProviderName = _context.Configuration.OperationGroupMapping.ContainsKey(operations.Key) ? _context.Configuration.OperationGroupMapping[operations.Key] : ConstructOperationProviderName(operations);
+            }
+        }
+
+        private string ConstructOperationProviderName(OperationGroup operations)
+        {
+
+            string? providerName = "";
+            var request = GetBestMethod(operations);
+            bool adding = false;
+            if (request != null)
+            {
+                foreach (var segment in GetPathSegments(request.Path))
+                {
+                    var asSplit = segment.Split('/', StringSplitOptions.RemoveEmptyEntries);
+                    if (asSplit?.Length > 1 && asSplit.First().Equals("providers"))
+                    {
+                        adding = true;
+                        providerName = segment.Substring("providers".Length + 2).TrimEnd('/');
+                    }
+                    else if (adding)
+                    {
+                        providerName += segment.TrimEnd('/');
+                    }
+                }
+            }
+            return providerName.TrimEnd('/');
+        }
+        private static List<string> GetPathSegments(string httpRequestUri)
+        {
+            List<string> seg = new List<string>();
+            string canidate = "";
+
+            foreach (var ch in httpRequestUri)
+            {
+                if (ch == '{')
+                {
+                    if (canidate != "" && canidate != "/")
+                    {
+                        seg.Add(canidate);
+                    }
+                    canidate = "";
+                }
+                else if (ch == '}')
+                {
+                    canidate = "";
+                }
+                else
+                {
+                    canidate += ch;
+                }
+            }
+            if (canidate != "" && canidate != "/")
+            {
+                seg.Add(canidate);
+            }
+            return seg;
+        }
+
+        private HttpRequest? GetBestMethod(OperationGroup operations)
+        {
+            HttpRequest? canidate = null;
+            foreach (var x in operations.Operations)
+            {
+                foreach (var serviceRequest in x.Requests)
+                {
+                    if (serviceRequest.Protocol.Http is HttpRequest httpRequest)
+                    {
+                        if (httpRequest.Method == HttpMethod.Put)
+                        {
+                            return httpRequest;
+                        }
+                        else if (httpRequest.Method == HttpMethod.Delete)
+                        {
+                            canidate = httpRequest;
```

Comment:
why do we not just return here?

Reply:
What if there is a PUT? 
 PUT->Delete->Patch is the order we want

Reply:
yes, but I believe the if order covers that, you are returning on first match and matching in the desired order

you can have 3 of these in the right order

if(httpRequest.Method == HttpMethod.Put)
    return httpRequest;

Reply:
yeah, but the operations can be in any order, 

so could get {delete, patch, put} in

(var operation in operationsGroup.Operations)


and we would return delete if we returned on the first entry in the operations list

hence, need check the rest to see if there is a put.